pkg/didcomm/protocol/presentproof/service.go

```diff
@@ -484,25 +489,42 @@ func (s *Service) deleteTransitionalPayload(id string) error {
 
 // Actions returns actions for the async usage.
 func (s *Service) Actions() ([]Action, error) {
-	records := s.store.Iterator(
-		fmt.Sprintf(transitionalPayloadKey, ""),
-		fmt.Sprintf(transitionalPayloadKey, storage.EndKeySuffix),
-	)
-	defer records.Release()
+	records, err := s.store.Query(transitionalPayloadKey)
+	if err != nil {
+		return nil, fmt.Errorf("failed to query store: %w", err)
+	}
+
+	defer func() {
+		errClose := records.Close()
+		if errClose != nil {
+			logger.Errorf("failed to close records iterator: %s", errClose.Error())
+		}
+	}()
 
 	var actions []Action
 
-	for records.Next() {
+	more, err := records.Next()
```

Comment:
I think the API's error-handling design can be improved significantly: https://dave.cheney.net/practical-go/presentations/qcon-china.html#_eliminate_error_handling_by_eliminating_errors

Reply:
I have about a half dozen storage implementations of the current API. Can you take a look at them and let me know what specifically you're thinking? I'm not really sure what action I should take based on that link.

Reply:
It's not a blocker. But the API's design could use improvements. Like the link says - stop returning so many errors that users have to handle.

Reply:
@llorllale I agree that the constant error returning is for sure a pain... The problem is that it's hard to get away from, since each storage implementation works differently. Some generate potential errors on `Next()`, some generate potential errors on `Value()`, some generate potential errors on several of those methods, etc... Apart from doing deferred error handling with an `Error()` method in the iterator interface (which is how the old interface worked), I'm not sure what else can be done. I was avoiding that type of deferred error handling since it made it very easy to swallow errors and skip checking them...

On a separate note - one of my requirements was to avoid breaking the storage interface multiple times, so if there's an improvement to be made to the interface then now is the time to do it.

Reply:
@DRK3 the general strategy is to just defer these errors and the user checks `itr.Error() != nil` after use.

`itr.Next()` should return false if there are no more records _or_ there is an error.

The whole thing will collapse to just:

```go
	records, err := s.store.Query(transitionalPayloadKey)
	if err != nil {
		return nil, fmt.Errorf("failed to query store: %w", err)
	}

        defer ....

        for records.Next() {
             record, err := records.Value()
             if err != nil {
                   // handle error
             }

             // handle record
        }

        if itr.Error() != nil {
             // handle error
        }
```

Reply:
I feel confident all implementations can behave this way.

Reply:
@llorllale Indeed I could do that. But then errors can be easily missed, since nothing forces a check on the `Error()` method. It seems wrong to make an error ignorable (hence why we have linters to check for that normally, but if we rely on the developer to call `Error()` then the linter can't enforce that)...

Reply:
@DRK3 I was responding to what seemed to be the meat of your counterargument:

> The problem is that it's hard to get away from, since each storage implementation works differently. Some generate potential errors on Next(), some generate potential errors on Value(), some generate potential errors on several of those methods, etc... Apart from doing deferred error handling with an Error() method in the iterator interface (which is how the old interface worked), I'm not sure what else can be done.

Regarding what is "preferable" vs what is not - that is opinion-based and you can run a poll if you wish. My opinion is it should be simplified not just because it looks cleaner, but also (important) - **it implies less test cases in our unit tests**, which translates into more productivity.

Reply:
So if the `Next()` call doesn't return an error, then by my count there's one less error check... that would be one fewer test then, I believe...not a huge improvement, but yes I agree that would be nicer.

In terms of productivity... I would argue that missing an error check (resulting in potential bugs) would be very bad for productivity... perhaps much worse than the pain of adding one more unit test.

In terms of running a poll - I'll leave that up to you to pursue, since the current design was the one already reviewed & approved by everyone months ago. I'm happy to change it if everyone prefers the tradeoffs associated with an explicit Error() method. Changing the current design will also require updating all the storage implementations... but if we're going to do it, I think it's now or never.

Reply:
> So if the Next() call doesn't return an error, then by my count there's one less error check... that would be one fewer test then, I believe.

There are 3 less errors that users have to handle in the snippet.

Reply:
+1 more if you consider the utility function for closing the iterator.

Reply:
@llorllale Yes, you're right, two less errors actually.